performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+ * in workloads and the largest difference in elements of T[ ] is one.  The
+ * number of values in the input stream generated by each processor will be
+ * related to the number of global table entries on each processor.
+ *
+ * The MPI-1 version of RandomAccess treats the potential instance where the
+ * number of processors is a power of two as a special case, because of the
+ * significant simplifications possible because processor location and local
+ * offset can be determined by applying masks to the input stream values.
+ * The non power of two case uses an integer division to determine the processor
+ * location.  The integer division will be more costly in terms of machine
+ * cycles to perform than the bit masking operations
+ *
+ * For additional information on the GUPS metric, the HPCchallenge RandomAccess
+ * Benchmark,and the rules to run RandomAccess or modify it to optimize
+ * performance -- see http://icl.cs.utk.edu/hpcc/
+ */
+
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <assert.h>
+#include <time.h>
+#include <sys/time.h>
+#include <string.h>
+#include <pthread.h>
+#include <stdint.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_tagged.h>
+#include <rdma/fi_atomic.h>
+
+#include "ct_utils.h"
+
+#define PRINT(stream, fmt, args...) \
+	do { \
+		fprintf(stream, fmt, ##args); \
+		fflush(stream); \
+	} while (0)
+#define VERBOSE(fmt, args...) PRINT(stdout,"rank %d: " fmt, rank, ##args)
+#define ERROR(fmt, args...) PRINT(stderr, fmt, ##args)
+#if 1 
+# define DEBUG(fmt, args...) ;
+#else
+# define DEBUG(fmt, args...) \
+	do { \
+		ERROR("[%s:%d] " fmt, __FILE__, __LINE__, ##args); \
+		fflush(stderr); \
+	} while (0)
+#endif
+
+#define GNIX_EP_RDM_PRIMARY_CAPS                                               \
+	(FI_MSG | FI_RMA | FI_TAGGED | FI_ATOMICS |                            \
+	 FI_DIRECTED_RECV | FI_READ |                                          \
+	 FI_WRITE | FI_SEND | FI_RECV | FI_REMOTE_READ | FI_REMOTE_WRITE)
+
+/* Program-level defines */
+#define MAX_ALIGNMENT 65536
+/* #define MAX_MSG_SIZE (1<<22) */
+#define MAX_MSG_SIZE (8*1024) /* Current GNI provider max send size */
+#define MYBUFSIZE (MAX_MSG_SIZE + MAX_ALIGNMENT)
+
+#define TEST_DESC "Libfabric Random Access Test"
+#define HEADER "# " TEST_DESC " \n"
+#ifndef FIELD_WIDTH
+#   define FIELD_WIDTH 20
+#endif
+#ifndef FLOAT_PRECISION
+#   define FLOAT_PRECISION 2
+#endif
+
+/* Random number generator */
+#define POLY 0x0000000000000007UL
+#define PERIOD 1317624576693539401L
+
+/* Define 64-bit types and corresponding format strings
+ * for printf() and constants
+ */
+#define FSTR64 "%ld"
+#define FSTRU64 "%lu"
+#define ZERO64B 0LL
+#define HEAD_RANK (rank == 0)
+#define UNLOCKED 0
+#define LOCKED(rank) (rank + 1)
+
+#define CTPM_SIMPLE_TYPE_FUNC(LABEL, type, op) \
```

Comment:
You didn't move this yet.  Or were you planning on doing it with another commit.